### PR TITLE
[Lagrangian.Model] Use base class Data instead of protected class member for constraint id

### PR DESCRIPTION
**Pull request by alxbilger**
related to https://github.com/SofaDefrost/SoftRobots/pull/283

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Review by EulalieCoevoet — APPROVED** [review]

**Comment by alxbilger** [comment]
[ci-build][with-all-tests]

File: Sofa/framework/Core/src/sofa/core/behavior/Constraint.inl [hunk 1]
```diff
@@ -69,6 +69,7 @@ void Constraint<DataTypes>::buildConstraintMatrix(const ConstraintParams* cParam
 {
     if (cParams)
     {
+        m_constraintIndex.setValue(cIndex);
```
Comment (alxbilger) [inline Sofa/framework/Core/src/sofa/core/behavior/Constraint.inl]: @EulalieCoevoet what do you think about setting the value here, at a higher level? All derived classes would then set the value.

Comment (EulalieCoevoet) [reply Sofa/framework/Core/src/sofa/core/behavior/Constraint.inl]: Yes this is a good idea. I'll do the same in SoftRobots (SoftRobotsConstraint).

**Edit**: Is it okay to hide the information that this value is only available after the call to buildConstraintMatrix()? The answer can be yes, I'm just wondering.

Comment (alxbilger) [reply Sofa/framework/Core/src/sofa/core/behavior/Constraint.inl]: you're right, setting the id could be done outside of the function `buildConstraintMatrix`. That is why I suggest https://github.com/sofa-framework/sofa/pull/4786.

This way, the id can be set from the outside using the `setConstraintId` function, that you can find here:

https://github.com/sofa-framework/sofa/blob/05a148594b72b495457fb2f1644edaa1c155925f/Sofa/framework/Simulation/Core/src/sofa/simulation/mechanicalvisitor/MechanicalBuildConstraintMatrix.cpp#L33

I suggest to pause this PR, the time that https://github.com/sofa-framework/sofa/pull/4786 is merged. Can you review it?

**Review by bakpaul — APPROVED** [review]

**Pull request merged**